slidingwindow.h

```diff
@@ -1838,73 +1839,60 @@ void ConvolutionInputGenerator_1D_dilated_dws(
-	const unsigned int num_of_inputs = IFMDim_x * multiplying_factor;
-	unsigned int inp = 0, index_write=0, index_read = 0, j = 0, internal_counter = 0;
-	for (unsigned int count_image = 0; count_image < numReps; count_image++) {
-		for (unsigned int i = 0; i < baseIter; i++) {
-			#pragma HLS PIPELINE II=1
-			if (inp < cycles_read_block) {// Initial buffer of ConvKernelDim lines
-				ap_uint<SIMD*Input_precision> inElem;
-				inElem = in.read();
-				inputBuf[inp] = inElem;
-				inp++;
-				index_write++;
-			}
-			else { // Read & write & update
-				// Read & write input buffer
-				if (inp < num_of_inputs){
-					if (inp < buffer_size || internal_counter >= buffer_size-multiplying_factor){
-						ap_uint<SIMD*Input_precision> inElem;
-						inElem = in.read();
-						index_write = index_write < buffer_size ? index_write : index_write - buffer_size;
-						inputBuf[index_write] = inElem;
-						inp++;
-						index_write++;
-						internal_counter = internal_counter < buffer_size-1 ? internal_counter+1 : 0;
-					}
-					else{
-						internal_counter++;
-					}
-				}
+	hls::stream<ap_uint<SIMD*Input_precision>> &in,
+	hls::stream<ap_uint<SIMD*Input_precision>> &out,
+	unsigned const  numReps,
+	R const &r
+) {
+	static_assert((IFMChannels % SIMD) == 0, "SIMD parallelism must divide the number of IFM channels");
+	static_assert(OFMDim_x == ((IFMDim_x-ConvKernelDim_x)/Stride_x+1), "Unexpected OFM dimension");
 
-				// Write output
-				ap_uint<SIMD*Input_precision> outElem = inputBuf[index_read];
-				out.write(outElem);
+	constexpr unsigned  SIMD_COUNT  = IFMChannels / SIMD;
+	constexpr unsigned  BUFFER_SIZE = (ConvKernelDim_x-1) * SIMD_COUNT;
+	constexpr unsigned  OUTPUT_SIZE = OFMDim_x * ConvKernelDim_x * SIMD_COUNT;
+	constexpr unsigned  NUM_OF_INPUTS = IFMDim_x*SIMD_COUNT;
+	constexpr unsigned  WINDOW_SIZE = ConvKernelDim_x*SIMD_COUNT;
 
-				// Update read index pointer
-				if (j < buffer_size-1){
-					index_read++;
-          			j++;
-				}
-				else{
-					index_read = index_read + (multiplying_factor+1);
-          			j=0;
-				}
-				index_read = index_read < buffer_size ? index_read : index_read - buffer_size;
+	ap_uint<SIMD*Input_precision>  buffer[BUFFER_SIZE];
+	memory_resource(buffer, r);
+
+	for(unsigned  count_image = 0; count_image < numReps; count_image++) {
+		unsigned  ocnt = (ConvKernelDim_x-1)*SIMD_COUNT+(Stride_x-1);
+		unsigned  wp = 0;
+		unsigned  rp = 0;
+		unsigned  offset = 0;
+		unsigned  inp_count = 0;
+		for(unsigned  i = 0; i < 1+OUTPUT_SIZE; i++) {
+#pragma HLS PIPELINE II=1
+			bool const  re = i > 0;
+			bool const  we = (i < WINDOW_SIZE) || (ocnt < SIMD_COUNT*Stride_x);
+			if(re) {
+				out.write(buffer[rp]);
+				if(++offset == WINDOW_SIZE){
+					offset = 0;
+					rp = rp+SIMD_COUNT*(Stride_x-1);
 				}
-		} // End base_iter
-	} // End count_image
-} // End generator
+				if(++rp >= BUFFER_SIZE)  rp -= BUFFER_SIZE;
+				if(++ocnt >= WINDOW_SIZE)  ocnt = 0;
```

Comment:
It looks like `ocnt` is incremented only in single steps. Testing for equality would, thus, be safe and is cheaper in HW.

Reply:
Thank you again, you're right. `ocnt` is indeed incremented only in single steps, but the initial assignment of `ocnt` can be greater than `WINDOW_SIZE`. I could make change the code in such a way to initialize `ocnt` either to be `(ConvKernelDim_x-1)*SIMD_COUNT+(Stride_x-1)` or 0 in case it exceeds `WINDOW_SIZE`, making the initial assignment more expensive, but the comparator can then be replaced by testing for equality.